src/clr-angular/forms/select/select.ts

```diff
@@ -4,24 +4,73 @@
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 
-import {AfterContentInit, Component, ContentChild, ElementRef, HostListener, ViewChild} from "@angular/core";
+import {isPlatformBrowser} from "@angular/common";
+import {
+    AfterContentInit,
+    Component,
+    ContentChild,
+    ElementRef,
+    HostListener,
+    OnDestroy,
+    PLATFORM_ID,
+    Renderer2,
+    ViewChild
+} from "@angular/core";
+import {Subscription} from "rxjs";
 
 import {POPOVER_HOST_ANCHOR} from "../../popover/common/popover-host-anchor.token";
 import {IfOpenService} from "../../utils/conditional/if-open.service";
 import {TAB} from "../../utils/key-codes/key-codes";
+
+import {ClrOption} from "./option";
 import {ClrOptions} from "./options";
+import {OptionSelectionService} from "./providers/option-selection.service";
+import {SelectDomAdapter} from "./utils/select-dom-adapter";
+import {SelectNoopDomAdapter} from "./utils/select-noop-dom-adapter";
+
+// Fixes build error
+// @dynamic (https://github.com/angular/angular/issues/19698#issuecomment-338340211)
+export const selectDomAdapterFactory = (platformId: Object) => {
+    if (isPlatformBrowser(platformId)) {
+        return new SelectDomAdapter();
+    } else {
+        return new SelectNoopDomAdapter();
+    }
+};
 
 @Component({
     selector: "clr-select",
     templateUrl: "select.html",
-    providers: [IfOpenService, {provide: POPOVER_HOST_ANCHOR, useExisting: ElementRef}],
+    providers: [
+        IfOpenService, {provide: POPOVER_HOST_ANCHOR, useExisting: ElementRef}, OptionSelectionService,
+        {provide: SelectDomAdapter, useFactory: selectDomAdapterFactory, deps: [PLATFORM_ID]}
+    ],
     host: {"[class.clr-select]": "true"}
 })
-export class ClrSelect implements AfterContentInit {
+export class ClrSelect<T> implements AfterContentInit, OnDestroy {
     @ViewChild("input") input: ElementRef;
     @ContentChild(ClrOptions) options: ClrOptions;
+    private subscription: Subscription;
 
-    constructor(private ifOpenService: IfOpenService) {}
+    constructor(private ifOpenService: IfOpenService, private optionSelectionService: OptionSelectionService<T>,
+                private renderer: Renderer2, private domAdapter: SelectDomAdapter) {
+        this.initializeSubscriptions();
+    }
+
+    private initializeSubscriptions(): void {
+        this.subscription = this.optionSelectionService.renderSelectionChanged.subscribe((option: ClrOption<T>) => {
+            this.renderSelection(option);
+        });
+    }
+
+    private renderSelection(selectedOption: ClrOption<T>): void {
+        if (this.input && selectedOption) {
+            this.domAdapter.clearChildren(this.input.nativeElement);
+            const clone: HTMLElement = this.domAdapter.cloneNode(selectedOption.elRef.nativeElement);
```

Comment:
Have we considered using a dynamic component creation based on the selected option's data? 

Also, if we clone the DOM element, we would need to also remove its old attributes and styles and add its styles to make it work in the input. We would need to manually handle classes/styles/events from now on.

I think the selected option needs its own component. That way, wouldn't it be more Angular appropriate and easy to scale once we add multi-select options?

@youdz what's your take on it?

Reply:
We have a "custom selected option template" as part of the MVP, but we want the API for the simple case to be just as simple as a native select, so we have to resort to the cloning trick in that case.

The selected option template directive will come in a future PR, but still before we ship.